rcl_action/include/rcl_action/goal_handle.h

```diff
@@ -0,0 +1,347 @@
+ *
+ * For C, a macro can be used (for example `example_interfaces/Fibonacci`):
+ *
+ * ```c
+ * #include <rosidl_generator_c/action_type_support_struct.h>
+ * #include <example_interfaces/action/fibonacci.h>
+ * const rosidl_action_type_support_t * ts =
+ *   ROSIDL_GET_ACTION_TYPE_SUPPORT(example_interfaces, Fibonacci);
+ * ```
+ *
+ * For C++, a template function is used:
+ *
+ * ```cpp
+ * #include <rosidl_generator_cpp/action_type_support.hpp>
+ * #include <example_interfaces/action/fibonacci.h>
+ * using rosidl_typesupport_cpp::get_action_type_support_handle;
+ * const rosidl_action_type_support_t * ts =
+ *   get_action_type_support_handle<example_interfaces::action::Fibonacci>();
+ * ```
+ *
+ * The rosidl_action_type_support_t object contains action type specific
+ * information used to send or take goals, results, and feedback.
+ *
+ * Expected usage:
+ *
+ * ```c
+ * #include <rcl/rcl.h>
+ * #include <rcl_action/rcl_action.h>
+ * #include <rosidl_generator_c/action_type_support_struct.h>
+ * #include <example_interfaces/action/fibonacci.h>
+ *
+ * // ... initialize node
+ * const rosidl_action_type_support_t * ts =
+ *   ROSIDL_GET_ACTION_TYPE_SUPPORT(example_interfaces, Fibonacci);
+ * rcl_action_server_t action_server = rcl_action_get_zero_initialized_server();
+ * rcl_action_server_options_t action_server_ops = rcl_action_server_get_default_options();
+ * ret = rcl_action_server_init(&action_server, &node, ts, "fibonacci", &action_server_ops);
+ * // ... error handling
+ * rcl_action_goal_handle_t goal_handle = rcl_action_get_zero_initialized_goal_handle();
+ * ret = rcl_action_goal_handle_init(&goal_handle, &action_server, ts);
+ * // ... error handling, and on shutdown do finalization:
+ * ret = rcl_action_goal_handle_fini(&goal_handle, &action_server);
+ * // ... error handling for rcl_goal_handle_fini()
+ * ret = rcl_action_server_fini(&action_server, &node);
+ * // ... error handling for rcl_action_server_fini()
+ * // ... finalize and error handling for node
+ * ```
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[out] goal_handle preallocated, zero-initialized, goal handle structure
+ *   to be initialized
+ * \param[in] action_server valid rcl action server
+ * \param[in] type_support type support object for the action's type
+ * \return `RCL_RET_OK` if goal_handle was initialized successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_GOAL_HANDLE_INVALID` if the goal handle is invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_BAD_ALLOC` if allocating memory failed, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_goal_handle_init(
```

Comment:
Is it a correct assumption this function is only used internally by `rcl_action` since the goal handle is returned by `rcl_action_send_goal_response`?

Reply:
Yes, that is correct.